lhotse/dataset/speech_recognition.py

```diff
@@ -57,3 +58,122 @@ def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
 
     def __len__(self) -> int:
         return len(self.cut_ids)
+
+
+class K2SpeechRecognitionDataset(Dataset):
+    """
+    .. warning::
+        This is just a draft of how Lhotse can interact with K2 - it's under active development.
+
+    The PyTorch Dataset for the speech recognition task using K2 library.
+    Each item in this dataset is a dict of:
+
+    .. code-block::
+
+        {
+            'features': (T x F) tensor,
+            'supervisions': List[Dict] -> [
+                {
+                    'text': string,
+                    'fsa': k2.Fsa,
+                    'start_frame': int,
+                    'end_frame': int
+                } (multiplied N times, for each of the N supervisions present in the Cut)
+            ],
+            'supervisions_mask': (T) tensor
+        }
+
+    The ``supervisions_mask`` field is a mask that specifies which frames are covered by a supervision
+    by assigning a value of 1 (in this case: segments with transcribed speech contents),
+    and which are not by asigning a value of 0 (in this case: padding, contextual noise,
+    or in general the acoustic context without transcription).
+    """
+
+    def __init__(self, cuts: CutSet):
+        import k2
+        super().__init__()
+        self.cuts = cuts
+        self.cut_ids = list(self.cuts.ids)
+
+        # We're creating a vocabulary of all characters present in the training CutSet.
+        # To do that we iterate through all the supervisions, and use chain.from_iterable
+        # to convert a sequence of their 'text' fields into a flat iterable of characters.
+
+        # Add special symbols first
+        self.vocab = ['<eps>', '<sil>', '<unk>'] + sorted(
+            # Add symbols present in the corpus
+            set(
+                chain.from_iterable(
+                    list(supervision.text)
+                    for cut in self.cuts
+                    for supervision in cut.supervisions
+                )
+            )
+        )
+        self.symbol_table = k2.SymbolTable(
+            _id2sym={i: tok for i, tok in enumerate(self.vocab)},
+            _sym2id={tok: i for i, tok in enumerate(self.vocab)}
+        )
+
+    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
+        import k2
+        cut_id = self.cut_ids[idx]
+        cut = self.cuts[cut_id]
+
+        features = torch.from_numpy(cut.load_features())
+        mask = torch.from_numpy(cut.supervisions_feature_mask())
+
+        sil_id = self.symbol_table.get('<sil>')
+        eps_id = self.symbol_table.get('<eps>')
+
+        supervision_fsas = []
+        for supervision in cut.supervisions:
+            # TODO(pzelasko): I am not sure how to attach the symbol table to the FSA object
+            #                 (or whether that's need at all?)
```

Comment:
Symbol tables are only for visualization. 
If you like, you could attach a symbol table to an FSA using the following code

```python
fsa.symbols = self.symbol_table
```